back-end/routes/recommendations.js

```diff
@@ -63,59 +63,114 @@ function calculateGlobalFeatureAverages(features) {
+            const totalMoodScore = (valenceScore + energyScore + instrumentScore) / 3;
             const danceScore = feature.danceability;
-            const totalScore = (valenceScore * moodWeight) + (danceScore * danceWeight) + (tempoScore * mixWeight)
+            const tempoScore = 1 - Math.abs(lastSongFeatures.tempo - feature.tempo) / lastSongFeatures.tempo;
+            const toneScore = tonalityScore(lastSongFeatures, feature);
+            const totalMixScore = (tempoScore + toneScore) / 2;
+            const totalScore = (totalMoodScore * moodWeight) + (danceScore * danceWeight) + (totalMixScore * mixWeight)
             scores.set(recommendId, totalScore);
         });
 
         const sortedScores = Array.from(scores).sort((a, b) => b[1] - a[1]);
-
-        let count = 1;
-        sortedScores.forEach(([id, diff]) => {
-            count++;
-        });
+        const sortedIds = sortedScores.map(entry => entry[0]);
+        return sortedIds;
     } catch (error) {
         console.error("Error in gradingAlgorithm:", error);
     }
 }
 
-const fetchRecommendations = async (songs, access_token) => {
+const fetchRecommendations = async (songs, accessToken, exploreWeight) => {
     const lastSongIds = songs.map(song => (song.uri).split(':').pop()).slice(-5);
     const commaSeparatedIds = lastSongIds.join(',');
+    const max_popularity = 100 - Math.round(50 * exploreWeight);
+
     try {
-        const response = await fetch(`https://api.spotify.com/v1/recommendations?limit=40&seed_tracks=${commaSeparatedIds}`, {
+        const response = await fetch(`https://api.spotify.com/v1/recommendations?limit=100&seed_tracks=${commaSeparatedIds}&max_popularity=${max_popularity}`, {
             headers: {
-                'Authorization': `Bearer ${access_token}`
+                'Authorization': `Bearer ${accessToken}`
             }
         });
+        if (response.statusCode !== 200) {
+            return (response.statusCode).json({ error: body });
+        }
         const recommendData = await response.json();
-        const recommendations = recommendData.tracks;
-        return recommendations;
+        const trackRecommendations = recommendData.tracks;
+        return trackRecommendations;
     } catch (error) {
         console.error(error);
     }
 };
 
+const fetchRankedTracks = async (songIds, accessToken) => {
+    const top20RecommendationIds = songIds.slice(0, 20);
+    const commaSeparatedIds = top20RecommendationIds.join(',');
+    try {
+        const response = await fetch(`https://api.spotify.com/v1/tracks?ids=${commaSeparatedIds}`, {
+            headers: {
+                'Authorization': `Bearer ${accessToken}`
+            }
+        });
+        const trackData = await response.json();
+        const tracks = trackData.tracks;
+        return tracks;
+    } catch (error) {
+        console.error(error);
```

Comment:
Please do not commit `console.error`s

Reply:
best practice is- use a logger to log your errors so you have a paper trail of errors that have occurred in your app

Reply:
Awesome, let's discuss this in our 1:1 tomorrow!